lib/AST/NameLookup.cpp

```diff
@@ -1598,7 +1598,29 @@ bool DeclContext::lookupQualified(ArrayRef<NominalTypeDecl *> typeDecls,
       // object initializers.
       bool visitSuperclass = true;
       if (member.getBaseName() == DeclBaseName::createConstructor()) {
-        if (classDecl->inheritsSuperclassInitializers())
+        // HACK: Don't check whether we can look into the superclass if we're
+        // already validating one of the class' constructors. This is necessary
+        // in order to break a cycle with associated type inference where
+        // computing initializer overrides can trigger associated type inference
+        // which performs this logic to check whether we can inherit
+        // initializers, which can check overrides.
+        // FIXME: We need to find a more principled way of breaking these
+        // associated type inference cycles.
+        auto isRecursiveValidation = [&]() -> bool {
```

Comment:
I'm hesitant to put a semantic cycle breaker into the middle of lookup (especially one that potentially makes lookup quadratic in the number of members). What's the full cycle here?

Reply:
Ideally associated type reason would have no reason to trigger implicit initializer synthesis, since there's nothing we can infer from them.

Reply:
@CodaFi The full cycle looks like:

<pre><code>TypeChecker::addImplicitConstructors
addImplicitInheritedConstructorsToClass
collectNonOveriddenSuperclassInits
<b>ConstructorDecl::getOverriddenDecl</b>
...
ValueDecl::getInterfaceType
...
TypeChecker::resolveTypeWitness
...
AssociatedTypeInference::solve
...
WitnessChecker::lookupValueWitnesses
...
DeclContext::lookupQualified
ClassDecl::inheritsSuperclassInitializers
...
collectNonOveriddenSuperclassInits
<b>ConstructorDecl::getOverriddenDecl</b></code></pre>

Reply:
@slavapestov What about cases like this?

```swift
protocol P {
  associatedtype X
  init(x: X)
}

struct S : P {
  var x: Int
}
```

Reply:
Constructor synthesis doesn’t need to compute the interface type.  You can remove all of that and suck it into the ParamDecl case for the interface type. You’ll need to run a lookup for the VarDecl corresponding to a given Param in the generated memberwise init.

Reply:
@CodaFi While that will likely help resolve rdar://54979757, I'm not sure it will help with cases like the above where associated type inference will need the interface type in order to attempt to infer the type witness.

To be clear though, the cycle that this PR is breaking is unrelated to the interface type computation of a synthesized initializer, it's to do with checking whether we inherit a superclass' convenience inits (in which case name lookup just looks into the superclass).

Reply:
It won't necessarily help there either since that cycle is also through witness matching.  Can this cycle be broken in witness matching?

Reply:
@CodaFi Hmm, I don't *think* so, as we need to be able to control when to stop looking into the superclass on encountering potential recursive validation. We still want name lookup to return other results for witnesses, including synthesized initializers.

Reply:
@CodaFi What would you like me to do with this? I know it's gross, but I'd rather not regress this for 5.2.